File: Solutions/Hashtable/219ContainsDuplicateII.py
class Solution:
    #Given an array of integers and an integer k, find out whether there there are 
    #two distinct indices i and j in the array such that nums[i] = nums[j] 
    #and the difference between i and j is at most k.

    # @param {integer[]} nums
    # @param {integer} k
    # @return {boolean}
    def containsNearbyDuplicate(self, nums, k):
        dic = {}
        for i,v in enumerate(nums):
            if dic.get(v) is None:
                dic[v] = [i]
            else:
                dic[v].append(i)
         
        
        for item in dic.values():
            if len(item) <=1 :
                continue
            else:
                if (item[1] - item[0]) <= k:
                    return True
        return False
if __name__ == "__main__":
    s = Solution()
    nums = [1,2,3,1,1,1,1,2,3,4,6,5,7,4,3,2,4,5]
    k = 1;
    print s.containsNearbyDuplicate(nums,k)
    

File: Solutions/Hashtable/219ContainsDuplicateII.cpp
#include<iostream>
#include<vector>
#include <cstdlib>
#include<iterator>
#include<map>
#include<unordered_map>
using namespace std;


/**
    Given an array of integers and an integer k, find out whether there 
    there are two distinct indices i and j in the array such that nums[i] = nums[j] 
    and the difference between i and j is at most k.

    
    happyNumber.cpp
*/

class Solution {

       public:
       
        bool containsNearbyDuplicate(vector<int>& nums, int k) {
            
            unordered_map<int,int> hmap;
            size_t n = nums.size();
            if(n <= 1)
                return 0;
            if(k <=0)
                return 0;
            hmap[nums[0]] = 0;
            for(size_t i =1;i< n ; ++i){
               int a = nums[i];
               
               if(hmap[a] == 0){//
                   if(a == nums[0]){
                       if(i-0 <= k ){
                           
                       // cout<<i<<"->"<<hmap[a]<<endl;
                           
                       return 1;
                       }
                   } 
                   hmap[a] = i;
                   
               }else{
                   if(i - hmap[a] <= k){
                       
                       // cout<<i<<"->"<<hmap[a]<<endl;
                        return 1;
                   }
                   else
                        hmap[a] = i;
                   
               }
               
               
            }//for 
            return 0;
      
        }

    
            
};




int main(){
        Solution s;
        
        vector<int> nums = {
            // 1,2,3,-1,5,4,3,-1,6,3
            -1,-1
            };
        int k=4;
        unordered_map<int,int> map1;
        map<int,int> mp;
        cout<<mp[0]<<endl;
        int a = map1[1];
        // cout<<a<<endl;
        // return 0;
        cin >> k;
        bool res = s.containsNearbyDuplicate(nums,k);
        cout<<res<<endl;
        return 0;
      

}
